### PR TITLE
ref(writes-limiter): Add WriteLimiterFactory

### PR DESCRIPTION
**Pull request by MeredithAnya**
Addressing https://github.com/getsentry/sentry/pull/37745#discussion_r949779591 in https://github.com/getsentry/sentry/pull/37745

Right the `check_writes_limits` does both checking the limits, but also initializes a new `RedisSlidingWindowRateLimiter`. This causes some complexity because we end up looking up the config in order to get the options stored on the config in order for the redis rate limiter to be initialized. 

Instead I've pulled the logic out of initializing the rate limiters into the `WritesLimiterFactory`, and `get_ratelimiter` will get the right rate limiter based on the config. This means that each `WritesLimiter` has it's own redis rate limiter, and the factory is in charge of mapping the namespace to the `WritesLimiter` instances. 

This means as we add parameters like `db_backend` to be used to get the config, we don't have to pass those along to the rate limiter.

### CODE REVIEW

**Comment by lynnagara** [comment]
The writes limiter factory feels like a bit of overkill to me. Its only job seems to be to store a mapping of namespace to limiter instances. But this is not really needed as each individual indexer backend doesn't need to look up rate limiters by key, it only ever needs a single rate limiter instance. And no one else needs to look it up except the indexer. Why not just construct the rate limiter within the indexer backend itself?

**Comment by lynnagara** [comment]
> This means as we add parameters like db_backend to be used to get the config, we don't have to pass those along to the rate limiter.

Why do we want this property? IMO the writes limiter just manages quotas for each namespace. I think it's better if it does not know anything about the multiple DB backends or metrics config or anything like that. Instead the dependency should go the other way - each DB backend instantiates the writes limiter passing in only the minimal configuration the limiter needs to get the quota and define the namespace for redis. The limiter knows nothing about the storage that it is in front of.

**Comment by MeredithAnya** [comment]
> The writes limiter factory feels like a bit of overkill to me. Its only job seems to be to store a mapping of namespace to limiter instances. But this is not really needed as each individual indexer backend doesn't need to look up rate limiters by key, it only ever needs a single rate limiter instance. And no one else needs to look it up except the indexer. Why not just construct the rate limiter within the indexer backend itself?

The redis options are based on the `use_case_id` (and later also the `db_backend` as well but that is basically static per indexer). The `use_case_id`, however, is currently passed through so we could have each backend keep a mapping of use case to rate limiter.

**Comment by MeredithAnya** [comment]
> > This means as we add parameters like db_backend to be used to get the config, we don't have to pass those along to the rate limiter.
> 
> Why do we want this property? IMO the writes limiter just manages quotas for each namespace. I think it's better if it does not know anything about the multiple DB backends or metrics config or anything like that. Instead the dependency should go the other way - each DB backend instantiates the writes limiter passing in only the minimal configuration the limiter needs to get the quota and define the namespace for redis. The limiter knows nothing about the storage that it is in front of.

I'm not sure I understand what you are saying here, my change is to avoid the `WritesLimiter` having to know about the `db_backend`. Whether or not there is a`WritesLimiterFactory`, the options are passed during initialization vs. before where the `WritesLimiter` class was using the parameters passed in through `check_rate_limits` to get the config to get the options... etc

File: src/sentry/sentry_metrics/indexer/ratelimiters.py [hunk 1]
```diff
@@ -103,23 +103,15 @@ def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
...
-
-    def _get_rate_limiter(
-        self, use_case_id: UseCaseKey, namespace: str
-    ) -> RedisSlidingWindowRateLimiter:
-        if namespace not in self.rate_limiters:
-            options = get_ingest_config(use_case_id).writes_limiter_cluster_options
-            self.rate_limiters[namespace] = RedisSlidingWindowRateLimiter(**options)
-
-        return self.rate_limiters[namespace]
+    def __init__(self, namespace: str, **options: Any) -> None:
```
Comment (lynnagara) [inline src/sentry/sentry_metrics/indexer/ratelimiters.py:112]: Can we properly type this? It's not particularly clear what `options` are supposed to be here.

**Comment by lynnagara** [comment]
> > The writes limiter factory feels like a bit of overkill to me. Its only job seems to be to store a mapping of namespace to limiter instances. But this is not really needed as each individual indexer backend doesn't need to look up rate limiters by key, it only ever needs a single rate limiter instance. And no one else needs to look it up except the indexer. Why not just construct the rate limiter within the indexer backend itself?
> 
> The redis options are based on the `use_case_id` (and later also the `db_backend` as well but that is basically static per indexer). The `use_case_id`, however, is currently passed through so we could have each backend keep a mapping of use case to rate limiter.

1. Why should the writes limiter ever need to know about `db_backend`? Shouldn't it just behave the same for ever backend (with a different namespace obviously so the counts don't collide)?
2. Would you mind clarifying the purpose of the factory, and why we are picking this option over just instantiating the limiter instance inside the backend which seems more straightforward as it is only used there anyway?

**Review by lynnagara — APPROVED** [review]
Discussed offline, LGTM.

**Pull request merged**